### PR TITLE
Adds support for disabling hash agg with hll functions on coordinator 

### PR DESCRIPTION
**Pull request by velioglu**
Checks whether the master query created by the Citus contains an hll aggregate function. We need to implement it in this way instead of using hooks provided by Postgres, since we create the query for the coordinator from scratch.

### CODE REVIEW

File: src/backend/distributed/planner/multi_master_planner.c [hunk 1]
```diff
@@ -313,8 +384,12 @@ BuildAggregatePlan(Query *masterQuery, Plan *subPlan)
 		 * in group and order by with aggregate operations.
 		 * see nodeAgg.c:build_pertrans_for_aggref(). In that case we use
 		 * sorted agg strategy, otherwise we use hash strategy.
+		 *
+		 * If the master query contains hll aggregate functions and the client set
+		 * hll.enable_hashagg to off, then we choose to use group aggregation.
 		 */
-		if (!enable_hashagg || !groupingIsHashable || hasDistinctAggregate)
+		if (!enable_hashagg || !groupingIsHashable || hasDistinctAggregate ||
+			UseGroupAggregateWithHLL(masterQuery))
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: I would rather see this next to  `groupingIsHashable` and `hasDistinctAggregate` calculations.

File: src/backend/distributed/planner/multi_master_planner.c [hunk 2]
```diff
@@ -244,6 +250,71 @@ BuildSelectStatement(Query *masterQuery, List *masterTargetList, CustomScan *rem
...
+ * QueryContainsAggregateWithHLL returns true if the query has an hll aggregate
+ * function in it's target list.
+ */
+static bool
+QueryContainsAggregateWithHLL(Query *query)
+{
+	List *varList = NIL;
+	ListCell *varCell = NULL;
+
+	varList = pull_var_clause((Node *) query->targetList, PVC_INCLUDE_AGGREGATES);
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: could you make sure hll functions in having clause are handled here

File: src/backend/distributed/planner/multi_master_planner.c [hunk 3]
```diff
@@ -244,6 +250,71 @@ BuildSelectStatement(Query *masterQuery, List *masterTargetList, CustomScan *rem
 }
 
 
+/*
+ * QueryContainsAggregateWithHLL returns true if the query has an hll aggregate
+ * function in it's target list.
+ */
+static bool
+QueryContainsAggregateWithHLL(Query *query)
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: Our function definition logic is that
1 - we declare the function
2 - we use the function 
3 - we provide the implementation of the function at the first available place.

Thus function definition order should look like
```
BuildAggregatePlan
HasDistinctAggregate
UseGroupAggregateWithHLL
QueryContainsAggregateWithHLL
```

File: src/backend/distributed/planner/multi_master_planner.c [hunk 4]
```diff
@@ -244,6 +250,71 @@ BuildSelectStatement(Query *masterQuery, List *masterTargetList, CustomScan *rem
...
+	const char *gucStrValue = NULL;
+
+	/* If HLL extension is not loaded, return false */
+	if (hllId == InvalidOid)
+	{
+		return false;
+	}
+
+	/* If HLL is loaded but related GUC is not set, return false */
+	gucStrValue = GetConfigOption(HLL_ENABLE_HASH_GUC_NAME, false, false);
```
Comment (mtuncer) [inline src/backend/distributed/planner/multi_master_planner.c]: Can we have hll extension defined in a different schema and still has the guc with hll prefix ?

Comment (marcocitus) [reply src/backend/distributed/planner/multi_master_planner.c]: Should be fine. GUC prefixes are unrelated to schemas.

File: src/include/distributed/multi_logical_optimizer.h [hunk 5]
```diff
@@ -38,6 +38,7 @@
 #define HLL_ADD_AGGREGATE_NAME "hll_add_agg"
 #define HLL_UNION_AGGREGATE_NAME "hll_union_agg"
 #define HLL_CARDINALITY_FUNC_NAME "hll_cardinality"
+#define HLL_ENABLE_HASH_GUC_NAME "hll.enable_hashagg"
```
Comment (byucesoy) [inline src/include/distributed/multi_logical_optimizer.h]: I think the name enable_hashagg can be confused with PostgreSQL's enable_hashagg, We wouldn't want to explain all users the difference between the enable_hashagg and hll.enable_hashagg. Can we name it something else; force_groupagg maybe?

File: src/backend/distributed/planner/multi_master_planner.c (hunk 1 above)
Comment (velioglu) [reply src/backend/distributed/planner/multi_master_planner.c]: Added in that way to not decrease the performance of aggregates without hll type

**Pull request merged**